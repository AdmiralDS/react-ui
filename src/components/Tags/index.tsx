import type { FC, KeyboardEvent, MouseEvent, ReactNode } from 'react';
import React, { Children, cloneElement, HTMLAttributes, isValidElement } from 'react';
import styled from 'styled-components';

const TAG_MARGIN = '8px';
const TagsWrapper = styled.div`
  display: flex;
  flex-flow: wrap;
  & > * {
    margin-right: ${TAG_MARGIN};
  }

  & > *:last-child {
    margin-right: 0;
  }
`;

type Kind = 'neutral' | 'green' | 'blue' | 'red' | 'orange';

export interface TagsProps extends HTMLAttributes<HTMLDivElement> {
  /** Ширина тэгов */
  width?: number | string;
  /** Тип тэгов */
  kind?: Kind;
  /** Обработчик клика по тэгам */
  onClick?: (event: MouseEvent<HTMLElement> | KeyboardEvent<HTMLElement>) => void;
}

export const Tags: FC<TagsProps> = ({ children, width, kind, onClick }) => {
  return (
    <TagsWrapper>
      {Children.map(children, (child: ReactNode) => {
        if (!isValidElement(child)) {
          return null;
        }
        return cloneElement(child, {
          onClick: child.props.onClick || onClick,
          width: child.props.width || width,
          kind: child.props.kind || kind,
          ...child.props,
        });
      })}
    </TagsWrapper>
  );
};
