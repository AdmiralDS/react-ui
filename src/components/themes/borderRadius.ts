import type { Shape } from '#src/components/themes/common';
export const ALL_BORDER_RADIUS_VALUES = [
  'Border radius 0',
  'Border radius 2',
  'Border radius 4',
  'Border radius 6',
  'Border radius 8',
  'Border radius 10',
] as const;

export function smallGroupBorderRadius(shape: Shape): string {
  switch (shape.borderRadiusKind) {
    case 'Border radius 0':
      return '0';
    case 'Border radius 2':
      return '2px';
    case 'Border radius 4':
      return '4px';
    case 'Border radius 6':
      return '4px';
    case 'Border radius 8':
      return '4px';
    case 'Border radius 10':
      return '4px';
    default:
      return '4px';
  }
}

export function mediumGroupBorderRadius(shape: Shape): string {
  switch (shape.borderRadiusKind) {
    case 'Border radius 0':
      return '0';
    case 'Border radius 2':
      return '2px';
    case 'Border radius 4':
      return '4px';
    case 'Border radius 6':
      return '6px';
    case 'Border radius 8':
      return '8px';
    case 'Border radius 10':
      return '10px';
    default:
      return '4px';
  }
}

export function largeGroupBorderRadius(shape: Shape): string {
  switch (shape.borderRadiusKind) {
    case 'Border radius 0':
      return '0';
    case 'Border radius 2':
      return '4px';
    case 'Border radius 4':
      return '8px';
    case 'Border radius 6':
      return '12px';
    case 'Border radius 8':
      return '16px';
    case 'Border radius 10':
      return '20px';
    default:
      return '8px';
  }
}
