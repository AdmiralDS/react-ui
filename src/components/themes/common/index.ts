import { Color } from '#src/components/themes/common/color';
import { ZIndex } from '#src/components/themes/common/zIndex';
import baseStyled, { ThemedStyledInterface } from 'styled-components';

export * from './color';

export interface Shadow {
  ClickableActive: string;
  ClickableDefault: string;
  ClickableHover: string;
  NonClickable: string;
}

export interface Theme {
  color: Color;
  shadow: Shadow;
  zIndex: ZIndex;
  name: 'light' | 'dark';
}

export const styled = baseStyled as ThemedStyledInterface<Theme>;
