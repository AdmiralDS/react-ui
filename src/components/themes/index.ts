export * from './light';
export * from './dark';
export type { Theme } from './common';
