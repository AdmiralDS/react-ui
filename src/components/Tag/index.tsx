import { forwardRef, useLayoutEffect, useRef, useState } from 'react';

import { refSetter } from '../common/utils/refSetter';
import { checkOverflow } from '../common/utils/checkOverflow';
import { Tooltip } from '#src/components/Tooltip';

import { Wrapper, TagCircle, Icon, Text, StatusIcon } from './style';

export type TagDimension = 'm' | 's';

/**
 * @deprecated Помечено как deprecated в версии 4.14.0, будет удалено в версии 9.х.х.
 * Взамен используйте альтернативные значения 'neutral' | 'success' | 'primary' | 'danger' | 'warning'
 */
export type DeprecatedTagKind = 'green' | 'blue' | 'red' | 'orange';

export type TagKind = 'neutral' | 'success' | 'primary' | 'danger' | 'warning' | DeprecatedTagKind;

export interface TagVisualProps {
  /** Тип тэга. Можно выбрать из предложенных вариантов, либо задать свои цвета для тэга.
   * В случае когда статус задается через статусную метку (кружок), свойство background отвечает за цвет статусной метки.
   * В случае когда статус задается через цвет фона и обводки, свойство background отвечает за цвет фона,
   * свойство border отвечает за цвет обводки, свойство backgroundHover отвечает за цвет фона при ховере в случае активного тэга.
   */
  kind?: TagKind | { background: string; border?: string; backgroundHover?: string };
  /** Отображение статуса через цвет обводки и фона. По умолчанию, при statusViaBackground = false, отображение статуса
   * происходит через цветную статусную метку (цветной кружок рядом с текстом)
   */
  statusViaBackground?: boolean;
  /** Скрыть обводку тэга (при условии, что статус отображается через цвет обводки и фона) */
  isBorderHidden?: boolean;
  /** Отображение иконки. Иконка отображается только по левому краю и при условии, что статус отображается через цвет обводки и фона */
  icon?: React.ReactNode;
}

export interface TagSizeProps {
  /** Высота тэга */
  dimension?: TagDimension;
  /** Ширина тэга */
  width?: number | string;
  /**
   * Позволяет рендерить компонент, используя другой тег HTML (https://styled-components.com/docs/api#as-polymorphic-prop).
   * В storybook в качестве примера приведены несколько возможных вариантов этого параметра (кроме них можно использовать любой другой HTML тег).
   */
  as?: React.ElementType;
}

export interface TagProps extends React.HTMLAttributes<HTMLButtonElement>, TagVisualProps, TagSizeProps {}

export interface TagInternalProps {
  /** Для внутреннего использования! Отображение иконки отрытия выпадающего меню */
  statusIcon?: React.ReactNode;
}

export const Tag = forwardRef<HTMLElement, TagProps & TagInternalProps>(
  (
    {
      children,
      kind = 'neutral',
      dimension = 'm',
      width,
      statusViaBackground = false,
      isBorderHidden = false,
      icon,
      statusIcon,
      onClick,
      ...props
    },
    ref,
  ) => {
    const wrapperRef = useRef<HTMLDivElement>(null);
    const textRef = useRef<HTMLElement>(null);
    const [overflow, setOverflow] = useState(false);
    const [tooltipVisible, setTooltipVisible] = useState(false);
    const background: TagKind | string =
      typeof kind === 'object' ? (kind.background ? kind.background : 'neutral') : (kind as TagKind);
    const border: TagKind | string = isBorderHidden
      ? 'transparent'
      : typeof kind === 'object'
        ? !!kind.background && !!kind.border
          ? kind.border
          : 'neutral'
        : (kind as TagKind);
    const backgroundHover: TagKind | string =
      typeof kind === 'object'
        ? kind.backgroundHover
          ? kind.backgroundHover
          : kind.background
            ? kind.background
            : 'neutral'
        : (kind as TagKind);

    useLayoutEffect(() => {
      const element = textRef.current;
      if (element && checkOverflow(element) !== overflow) {
        setOverflow(checkOverflow(element));
      }
    }, [tooltipVisible, setOverflow]);

    useLayoutEffect(() => {
      function show() {
        setTooltipVisible(true);
      }
      function hide() {
        setTooltipVisible(false);
      }
      const wrapper = wrapperRef.current;
      if (wrapper) {
        wrapper.addEventListener('mouseenter', show);
        wrapper.addEventListener('mouseleave', hide);
        wrapper.addEventListener('focus', show);
        wrapper.addEventListener('blur', hide);
        return () => {
          wrapper.removeEventListener('mouseenter', show);
          wrapper.removeEventListener('mouseleave', hide);
          wrapper.removeEventListener('focus', show);
          wrapper.removeEventListener('blur', hide);
        };
      }
    }, [setTooltipVisible]);

    return (
      <>
        <Wrapper
          ref={refSetter(ref, wrapperRef)}
          $width={width}
          onClick={onClick}
          $clickable={!!onClick}
          $statusViaBackground={statusViaBackground}
          $border={border}
          $background={background}
          $backgroundHover={backgroundHover}
          $dimension={dimension}
          type="button"
          {...props}
        >
          {background !== 'neutral' && !statusViaBackground && <TagCircle $background={background} />}
          {statusViaBackground && icon && <Icon>{icon}</Icon>}
          {children && <Text ref={textRef}>{children}</Text>}
          {statusIcon && <StatusIcon>{statusIcon}</StatusIcon>}
        </Wrapper>
        {tooltipVisible && overflow && <Tooltip targetElement={wrapperRef.current} renderContent={() => children} />}
      </>
    );
  },
);

Tag.displayName = 'Tag';

export { TagCircle } from './style';
