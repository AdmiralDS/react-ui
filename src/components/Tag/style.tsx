import styled, { css } from 'styled-components';
import { typography } from '#src/components/Typography';
import { smallGroupBorderRadius } from '#src/components/themes/borderRadius';
import type { TagKind, TagDimension } from '#src/components/Tag';

const TAG_HEIGHT_S = 20;
const TAG_HEIGHT_M = 24;
const TAG_PADDING_TOP_S = 1;
const TAG_PADDING_TOP_M = 3;
const TAG_PADDING_LEFT_S = 5;
const TAG_PADDING_LEFT_M = 7;

const circleBackground = css<{ $background: TagKind | string }>`
  background: ${({ $background, theme }) => {
    switch ($background) {
      case 'success':
      case 'green':
        return `var(--admiral-color-Success_Success50Main, ${theme.color['Success/Success 50 Main']})`;
      case 'primary':
      case 'blue':
        return `var(--admiral-color-Primary_Primary60Main, ${theme.color['Primary/Primary 60 Main']})`;
      case 'danger':
      case 'red':
        return `var(--admiral-color-Error_Error60Main, ${theme.color['Error/Error 60 Main']})`;
      case 'warning':
      case 'orange':
        return `var(--admiral-color-Warning_Warning50Main, ${theme.color['Warning/Warning 50 Main']})`;
      default:
        return $background;
    }
  }};
`;

const wrapperBackground = css<{ $background: TagKind | string }>`
  background: ${({ $background, theme }) => {
    switch ($background) {
      case 'success':
      case 'green':
        return `var(--admiral-color-Success_Success10, ${theme.color['Success/Success 10']})`;
      case 'primary':
      case 'blue':
        return `var(--admiral-color-Primary_Primary10, ${theme.color['Primary/Primary 10']})`;
      case 'danger':
      case 'red':
        return `var(--admiral-color-Error_Error10, ${theme.color['Error/Error 10']})`;
      case 'warning':
      case 'orange':
        return `var(--admiral-color-Warning_Warning10, ${theme.color['Warning/Warning 10']})`;
      case 'neutral':
        return `var(--admiral-color-Opacity_Neutral8, ${theme.color['Opacity/Neutral 8']})`;
      default:
        return $background || `var(--admiral-color-Opacity_Neutral8, ${theme.color['Opacity/Neutral 8']})`;
    }
  }};
`;

const wrapperBorder = css<{ $border: TagKind | string }>`
  border: 1px solid
    ${({ $border, theme }) => {
      switch ($border) {
        case 'success':
        case 'green':
          return `var(--admiral-color-Success_Success40, ${theme.color['Success/Success 40']})`;
        case 'primary':
        case 'blue':
          return `var(--admiral-color-Primary_Primary50, ${theme.color['Primary/Primary 50']})`;
        case 'danger':
        case 'red':
          return `var(--admiral-color-Error_Error50, ${theme.color['Error/Error 50']})`;
        case 'warning':
        case 'orange':
          return `var(--admiral-color-Warning_Warning40, ${theme.color['Warning/Warning 40']})`;
        case 'neutral':
          return `var(--admiral-color-Neutral_Neutral40, ${theme.color['Neutral/Neutral 40']})`;
        default:
          return $border || `var(--admiral-color-Neutral_Neutral40, ${theme.color['Neutral/Neutral 40']})`;
      }
    }};
`;

const wrapperHover = css<{ $backgroundHover: TagKind | string }>`
  background: ${({ $backgroundHover, theme }) => {
    switch ($backgroundHover) {
      case 'success':
      case 'green':
        return `var(--admiral-color-Success_Success20, ${theme.color['Success/Success 20']})`;
      case 'primary':
      case 'blue':
        return `var(--admiral-color-Primary_Primary20, ${theme.color['Primary/Primary 20']})`;
      case 'danger':
      case 'red':
        return `var(--admiral-color-Error_Error20, ${theme.color['Error/Error 20']})`;
      case 'warning':
      case 'orange':
        return `var(--admiral-color-Warning_Warning20, ${theme.color['Warning/Warning 20']})`;
      case 'neutral':
        return `var(--admiral-color-Opacity_Neutral12, ${theme.color['Opacity/Neutral 12']})`;
      default:
        return $backgroundHover || `var(--admiral-color-Opacity_Neutral12, ${theme.color['Opacity/Neutral 12']})`;
    }
  }};
`;

export const Wrapper = styled.button<{
  $dimension?: TagDimension;
  $width?: number | string;
  $clickable: boolean;
  $statusViaBackground?: boolean;
  $border: TagKind | string;
  $background: TagKind | string;
  $backgroundHover: TagKind | string;
}>`
  position: relative;
  box-sizing: border-box;
  max-width: 100%;

  ${({ $dimension }) => `
    height: ${$dimension === 's' ? TAG_HEIGHT_S : TAG_HEIGHT_M}px;
    padding: ${$dimension === 's' ? TAG_PADDING_TOP_S : TAG_PADDING_TOP_M}px ${
      $dimension === 's' ? TAG_PADDING_LEFT_S : TAG_PADDING_LEFT_M
    }px;
  `}

  ${({ $width }) => $width && `width: ${typeof $width === 'number' ? `${$width}px` : $width};`}
  border-radius: var(--admiral-border-radius-Small, ${(p) => smallGroupBorderRadius(p.theme.shape)});
  ${({ $statusViaBackground, theme }) =>
    $statusViaBackground
      ? wrapperBackground
      : `background: var(--admiral-color-Opacity_Neutral8, ${theme.color['Opacity/Neutral 8']});`}
  ${({ $statusViaBackground }) => ($statusViaBackground ? wrapperBorder : `border: 1px solid transparent;`)}

  display: inline-flex;
  align-items: center;
  cursor: ${({ $clickable }) => ($clickable ? 'pointer' : 'default')};

  &:hover,
  &:active {
    ${({ $statusViaBackground, $clickable }) =>
      $clickable && !$statusViaBackground ? `border: 1px solid transparent;` : ''}
    ${({ $statusViaBackground, theme, $clickable }) =>
      $clickable
        ? $statusViaBackground
          ? wrapperHover
          : `background: var(--admiral-color-Opacity_Neutral12, ${theme.color['Opacity/Neutral 12']});`
        : ''}
  }

  &:focus-visible {
    outline: none;

    &:before {
      content: '';
      position: absolute;
      top: -4px;
      bottom: -4px;
      left: -4px;
      right: -4px;
      border-radius: var(--admiral-border-radius-Small, ${(p) => smallGroupBorderRadius(p.theme.shape)});
      border: 2px solid var(--admiral-color-Primary_Primary60Main, ${(p) => p.theme.color['Primary/Primary 60 Main']});
    }
  }
`;

export const Text = styled.span`
  color: var(--admiral-color-Neutral_Neutral90, ${(p) => p.theme.color['Neutral/Neutral 90']});
  user-select: none;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;

  ${typography['Caption/Caption 1']}
`;

export const TagCircle = styled.div<{ $background: TagKind | string }>`
  display: flex;
  flex-shrink: 0;
  width: 8px;
  height: 8px;
  border-radius: 50%;
  margin-right: 6px;
  ${circleBackground}
`;

export const Icon = styled.div`
  display: flex;
  flex-shrink: 0;
  width: 16px;
  height: 16px;
  margin-right: 4px;

  & *[fill^='#'] {
    fill: var(--admiral-color-Neutral_Neutral50, ${(p) => p.theme.color['Neutral/Neutral 50']});
  }
`;

export const StatusIcon = styled.div`
  display: flex;
  flex-shrink: 0;
  width: 16px;
  height: 16px;
  margin-left: 4px;

  & *[fill^='#'] {
    fill: var(--admiral-color-Neutral_Neutral50, ${(p) => p.theme.color['Neutral/Neutral 50']});
  }

  &:hover {
    & *[fill^='#'] {
      fill: var(--admiral-color-Neutral_Neutral50, ${(p) => p.theme.color['Neutral/Neutral 50']});
    }
  }
`;
