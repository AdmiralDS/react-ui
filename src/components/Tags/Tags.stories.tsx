import React from 'react';
import styled from 'styled-components';
import { ComponentMeta, ComponentStory } from '@storybook/react';
import { withDesign } from 'storybook-addon-designs';

import { Tag } from '../Tag';
import { Tags } from '../Tags';

const Separator = styled.div`
  height: 20px;
`;

const Desc = styled.div`
  font-family: 'VTB Group UI';
  font-size: 16px;
  line-height: 24px;
`;

const Description = () => (
  <Desc>
    Тег — это метка, размечающая и каталогизирующая информацию для облегчения процесса поиска. При нажатии на тэг
    загружаются все элементы, имеющие эту метку. Тэги можно использовать группами. Горизонтальные и вертикальные отступы
    между соседними тэгами равны 8px.
    <Separator />
    Для каждого тэга можно отдельно задать width, kind и onClick, либо можно задать единые width, kind и onClick через
    Tags.
  </Desc>
);

export default {
  title: 'Example/Tags',
  decorators: [withDesign],
  component: Tags,
  parameters: {
    componentSubtitle: <Description />,
    layout: 'centered',
    design: [
      {
        type: 'figma',
        url: 'https://www.figma.com/file/CC0WL5u9TPtZpyLbbAGFGt/Admiral-2.0-UI-Kit?node-id=37%3A18530',
      },
      {
        type: 'figma',
        url: 'https://www.figma.com/file/CC0WL5u9TPtZpyLbbAGFGt/Admiral-2.0-UI-Kit?node-id=37%3A18660',
      },
    ],
  },
  argTypes: {
    kind: {
      options: ['neutral', 'blue', 'green', 'red', 'orange'],
      control: { type: 'radio' },
    },
    width: {
      control: { type: 'text' },
    },
  },
} as ComponentMeta<typeof Tags>;

const Template1: ComponentStory<typeof Tags> = () => {
  return (
    <Tags width={50} kind="green" onClick={(event) => console.log(`click tag with id: ${event.currentTarget.id}`)}>
      <Tag id="1">Neutral</Tag>
      <Tag id="2" kind="green">
        Green
      </Tag>
      <Tag id="3" kind="blue">
        Blue
      </Tag>
      <Tag id="4" kind="red" width="auto" onClick={() => console.log('click red tag')}>
        Red
      </Tag>
      <Tag id="5" kind="orange">
        Orange
      </Tag>
    </Tags>
  );
};

export const TagStatus = Template1.bind({});
TagStatus.args = {};
TagStatus.storyName = 'Tags';
