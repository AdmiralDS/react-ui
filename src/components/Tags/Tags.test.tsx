import React from 'react';
import { fireEvent, render } from '@testing-library/react';
import { ThemeProvider } from 'styled-components';

import { LIGHT_THEME } from '../themes';
import { Tags } from '../Tags';
import { Tag } from '../Tag';

describe('Tags', () => {
  const clickProps = {
    onClickMain: jest.fn(),
    onClickRed: jest.fn(),
  };
  const Component = () => (
    <ThemeProvider theme={LIGHT_THEME}>
      <Tags width={50} kind="green" onClick={clickProps.onClickMain}>
        <Tag data-testid="1">Neutral</Tag>
        <Tag data-testid="2" kind="red" width="auto" onClick={clickProps.onClickRed}>
          Red
        </Tag>
        <Tag data-testid="3">Orange</Tag>
      </Tags>
    </ThemeProvider>
  );
  it('should render default component', () => {
    const wrapper = render(<Component />);
    expect(wrapper).toMatchSnapshot();
  });

  it('should call onClickMain when user clicks on first tag', () => {
    const wrapper = render(<Component />);
    fireEvent.click(wrapper.getByTestId('1'));
    const { onClickMain } = clickProps;
    expect(onClickMain).toBeCalledTimes(1);
  });

  it('should call onClickred when user clicks on second tag', () => {
    const wrapper = render(<Component />);
    fireEvent.click(wrapper.getByTestId('2'));
    const { onClickRed } = clickProps;
    expect(onClickRed).toBeCalledTimes(1);
  });
});
