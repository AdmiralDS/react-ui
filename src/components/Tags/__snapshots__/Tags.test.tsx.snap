// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tags should render default component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: wrap;
  -ms-flex-flow: wrap;
  flex-flow: wrap;
}

.c0 > * {
  margin-right: 8px;
}

.c0 > *:last-child {
  margin-right: 0;
}

.c1 {
  position: relative;
  box-sizing: border-box;
  height: 24px;
  width: 50px;
  margin-top: 8px;
  padding: 3px 7px;
  border-radius: 4px;
  background: #E2E5E9;
  border: 1px solid #E2E5E9;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
}

.c1:hover,
.c1:active {
  background: #D3D8DE;
}

.c1:focus {
  outline: none;
}

.c1:focus:before {
  content: '';
  position: absolute;
  top: -4px;
  bottom: -4px;
  left: -4px;
  right: -4px;
  border-radius: 6px;
  border: 2px solid #0A54C2;
}

.c4 {
  position: relative;
  box-sizing: border-box;
  height: 24px;
  width: auto;
  margin-top: 8px;
  padding: 3px 7px;
  border-radius: 4px;
  background: #E2E5E9;
  border: 1px solid #E2E5E9;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
}

.c4:hover,
.c4:active {
  background: #D3D8DE;
}

.c4:focus {
  outline: none;
}

.c4:focus:before {
  content: '';
  position: absolute;
  top: -4px;
  bottom: -4px;
  left: -4px;
  right: -4px;
  border-radius: 6px;
  border: 2px solid #0A54C2;
}

.c3 {
  color: #2B313B;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-family: 'VTB Group UI',sans-serif;
  font-style: normal;
  font-weight: normal;
  font-size: 12px;
  line-height: 16px;
  font-feature-settings: 'tnum' on,'lnum' on;
  color: #2B313B;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 8px;
  height: 8px;
  border-radius: 50%;
  margin-right: 6px;
  background: #2ABB5B;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 8px;
  height: 8px;
  border-radius: 50%;
  margin-right: 6px;
  background: #DA0B20;
}

<body>
    <div>
      <div
        class="c0"
      >
        <button
          class="c1"
          data-testid="1"
          width="50"
        >
          <div
            class="c2"
          />
          <span
            class="c3"
          >
            Neutral
          </span>
        </button>
        <button
          class="c4"
          data-testid="2"
          width="auto"
        >
          <div
            class="c5"
          />
          <span
            class="c3"
          >
            Red
          </span>
        </button>
        <button
          class="c1"
          data-testid="3"
          width="50"
        >
          <div
            class="c2"
          />
          <span
            class="c3"
          >
            Orange
          </span>
        </button>
      </div>
    </div>
  </body>,
  "container": .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-flow: wrap;
  -ms-flex-flow: wrap;
  flex-flow: wrap;
}

.c0 > * {
  margin-right: 8px;
}

.c0 > *:last-child {
  margin-right: 0;
}

.c1 {
  position: relative;
  box-sizing: border-box;
  height: 24px;
  width: 50px;
  margin-top: 8px;
  padding: 3px 7px;
  border-radius: 4px;
  background: #E2E5E9;
  border: 1px solid #E2E5E9;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
}

.c1:hover,
.c1:active {
  background: #D3D8DE;
}

.c1:focus {
  outline: none;
}

.c1:focus:before {
  content: '';
  position: absolute;
  top: -4px;
  bottom: -4px;
  left: -4px;
  right: -4px;
  border-radius: 6px;
  border: 2px solid #0A54C2;
}

.c4 {
  position: relative;
  box-sizing: border-box;
  height: 24px;
  width: auto;
  margin-top: 8px;
  padding: 3px 7px;
  border-radius: 4px;
  background: #E2E5E9;
  border: 1px solid #E2E5E9;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  cursor: pointer;
}

.c4:hover,
.c4:active {
  background: #D3D8DE;
}

.c4:focus {
  outline: none;
}

.c4:focus:before {
  content: '';
  position: absolute;
  top: -4px;
  bottom: -4px;
  left: -4px;
  right: -4px;
  border-radius: 6px;
  border: 2px solid #0A54C2;
}

.c3 {
  color: #2B313B;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-family: 'VTB Group UI',sans-serif;
  font-style: normal;
  font-weight: normal;
  font-size: 12px;
  line-height: 16px;
  font-feature-settings: 'tnum' on,'lnum' on;
  color: #2B313B;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 8px;
  height: 8px;
  border-radius: 50%;
  margin-right: 6px;
  background: #2ABB5B;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  width: 8px;
  height: 8px;
  border-radius: 50%;
  margin-right: 6px;
  background: #DA0B20;
}

<div>
    <div
      class="c0"
    >
      <button
        class="c1"
        data-testid="1"
        width="50"
      >
        <div
          class="c2"
        />
        <span
          class="c3"
        >
          Neutral
        </span>
      </button>
      <button
        class="c4"
        data-testid="2"
        width="auto"
      >
        <div
          class="c5"
        />
        <span
          class="c3"
        >
          Red
        </span>
      </button>
      <button
        class="c1"
        data-testid="3"
        width="50"
      >
        <div
          class="c2"
        />
        <span
          class="c3"
        >
          Orange
        </span>
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
